---
title: "micro-RNA Vesicles"
output:
  html_notebook: default
---
**Устанавливаем необходимые библиотки**
```{r}
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(clusterProfiler)
library(biomaRt)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(GenomicRanges)
library(msigdbr)
library(multiMiR)
library(miRBaseConverter)
library(enrichplot)
library(vsn)
library(rvest)
library(patchwork)
library(dbplyr)
```
## Импортируем данные
```{r}
coldata <- read_tsv("data/phenotable.tsv", show_col_types = FALSE)
rownames(coldata) <- coldata$sample
coldata
```

```{r}
counts <- read.csv("data/miR.Counts.csv", header = TRUE, sep = ",")
counts <- column_to_rownames(counts, var = "miRNA")
head(counts)
```
```{r}
colnames(counts) <- gsub("^X", "", colnames(counts))

counts_samples <- colnames(counts)
phenotable_samples <- coldata$sample
common_samples <- intersect(counts_samples, phenotable_samples)

counts <- counts[, c(counts$miRNA, common_samples)]  
counts <- counts[, rownames(coldata)] #ранжирую по колонки в counts так же как и названия строк в coldata
head(counts)
```
```{r}
anno <- read.csv("data/annotation.report.csv", header = TRUE, sep = ",")
anno$Sample.name.s. <- gsub("-", ".", anno$Sample.name.s.)
anno <- anno[, -c(2:5, 7, 15)]


common_samples <- intersect(anno$Sample.name.s., coldata$sample)

anno <- anno[anno$Sample.name.s. %in% common_samples, ]
anno <- anno[match(rownames(coldata), anno$Sample.name.s.), ] #ранжирую по колонки в counts так же как и названия строк в coldata
anno
```
## Весь датасет
```{r}
anno_long <- anno %>%
  pivot_longer(cols = -Sample.name.s., names_to = "RNA_Type", values_to = "Count")

plt <- ggplot(anno_long, aes(x = Sample.name.s., y = Count, fill = RNA_Type)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Sample", y = "Read Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_brewer(palette = "Set3")  # Красивые цвета

print(plt)
ggsave("./pictures_transpl/transpl_barplot_alldataset_no_normalised.tiff", plot = plt, width = 8, height = 6, dpi = 300,  bg = "white")
```

```{r}
anno_long <- anno %>%
  rowwise() %>%
  mutate(across(-Sample.name.s., ~ . / sum(c_across(-Sample.name.s.)))) %>% 
  ungroup() %>%
  pivot_longer(cols = -Sample.name.s., names_to = "RNA_Type", values_to = "Proportion")

plt <- ggplot(anno_long, aes(x = Sample.name.s., y = Proportion, fill = RNA_Type)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(x = "Sample", y = "Proportion") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_brewer(palette = "Set3")

plt
ggsave("./pictures_transpl/transpl_barplot_alldataset_normalised.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
```{r}
coldata$condition <- relevel(factor(coldata$condition), ref = "no_complications")
modelMatrix <- model.matrix(~condition, coldata)
modelMatrix
qr(modelMatrix)$rank  # ранг матрицы
ncol(modelMatrix) 
```


```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ condition)
dds$condition <- relevel(dds$condition, ref = "no_complications")
dds
```
```{r}
dim(dds)
smallestGroupSize <- 15
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
dim(dds)
```
###  Run Differential Expression Analysis ###
```{r}
dds <- DESeq(dds, fitType = "parametric")
dds
```

```{r}
plotDispEsts(dds)
```
```{r}
raw_counts <- counts(dds, normalized = FALSE)
normalized_counts <- counts(dds, normalized = TRUE)

df <- data.frame(
  Sample = rep(colnames(dds), 2),
  Counts = c(colSums(raw_counts), colSums(normalized_counts)),
  Type = rep(c("Raw", "Normalized"), each = ncol(dds))
)

plt <- ggplot(df, aes(x = Sample, y = Counts, fill = Type)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Counts before and after normalization", x = "Sample", y = "Total Counts") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

plt
ggsave("./pictures_transpl/transpl_Counts before and after normalization.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
rlog трансформация
```{r}
rlt <- rlog(dds)  #rlog Transformation
meanSdPlot(assay(rlt)) 

```
```{r}
vsd <- varianceStabilizingTransformation(dds, blind=FALSE) 
meanSdPlot(assay(vsd)) #показывает, как изменяется стандартное отклонение в зависимости от среднего значения экспрессии
```

** PCA plot **
```{r}
pcaData <- plotPCA(rlt, intgroup=c("condition"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

coldata$sample <- gsub("_.*", "", coldata$sample)
pcaData$sample <- coldata$sample

plt <- ggplot(pcaData, aes(PC1, PC2, color = condition)) +
  geom_text(aes(label=sample), size=3, vjust=1.5) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "%")) +
  ylab(paste0("PC2: ", percentVar[2], "%")) + 
  coord_fixed() +
  theme_bw() +
  scale_color_brewer(palette = "Set2")
plt
ggsave("./pictures_transpl/transpl_PCA plot.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```

**Plot a heatmap of 50 most expressed genes**
Этот heatmap отражает уровни экспрессии генов, а не разницу между группами.
Цвета не означают up- или down-регуляцию в сравнении с контрольной группой, потому что heatmap показывает абсолютные значения экспрессии, а не fold change!
```{r}
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:50]
df <- as.data.frame(colData(dds)$condition)
colnames(df) <- "condition"
rownames(df) <- colnames(counts(dds))
plt <- pheatmap(assay(rlt)[select,], 
         cluster_rows = TRUE, 
         show_rownames = TRUE, 
         cluster_cols = TRUE, 
         annotation_col = df,
         fontsize_row = 6) 

plt
ggsave("./pictures_transpl/transpl_Plot a heatmap of 50 most expressed genes.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
**Plot of the distance between samples heatmap**
Расчет расстояний между образцами
	•	Обычно используется евклидово расстояние (по умолчанию в DESeq2).
	•	Оно вычисляется по нормализованным данным экспрессии (rlog() или vst()).
	•	Чем меньше расстояние — тем более похожи образцы.
```{r}
sampleDists <- dist(t(assay(rlt)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rlt$condition)
colnames(sampleDistMatrix) <- paste(rlt$condition)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

plt <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         color = colors)

plt
ggsave("./pictures_transpl/transpl_Plot of the distance between samples heatmap.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
```{r}
res_humoral <- results(dds, contrast=c("condition", "no_complications", "humoral"))
res_humoral
```
**MA plot**
Фильтрация точек с низким средним экспрессированием (по baseMean).
	•	Обычно отсекаются baseMean < 1.
	2.	Определение значимых генов (синие точки):
	•	Используется критерий padj < 0.1 по умолчанию, а не < 0.05! 
```{r}
tiff("./pictures_transpl/transpl_PlotMA_standart_padj_0.05_humoral.tiff", 
     width = 8, height = 6, units = "in", res = 300, bg = "white")
plotMA(res_humoral, alpha = 0.05, ylim = c(-8, 8)) 
dev.off()
plotMA(res_humoral, alpha = 0.05, ylim = c(-8, 8)) 
```
**Кастомный MA plot**
```{r}
res_df <- res_humoral %>%
  as.data.frame() %>%
  mutate(color = case_when( 
    padj < 0.05  ~ "padj < 0.05",   
    pvalue < 0.05  ~ "pvalue < 0.05", 
    TRUE ~ "All"
  ))

plt <- ggplot(res_df, aes(x = baseMean, y = log2FoldChange, color = color)) +
  geom_point(alpha = 0.7, size = 1) +
  geom_hline(yintercept = 0, linetype = "solid", color = "gray40", size = 1.5) +
  scale_color_manual(values = c("All" = "gray70", 
                                "pvalue < 0.05" = "blue", 
                                "padj < 0.05" = "red")) +
  scale_x_log10(labels = scales::scientific) + 
  theme_minimal() +
  labs(x = "mean of normalized counts", 
       y = "log fold change", 
       color = NULL)

plt
ggsave("./pictures_transpl/transpl_Сustom MAplot_humoral.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```

**Значимые результаты**

58 генов (20%) имеют низкие уровни экспрессии и фильтруются из анализа.
independent filtering — процедура, которая исключает гены с низкими значениями для увеличения статистической мощности.
```{r}
signres_humoral <- results(dds, contrast=c("condition", "no_complications", "humoral"), alpha=0.05) 
summary(signres_humoral)
```
Let's arranged it by log2FoldChange:
```{r}
order_indices <- order(-res_humoral$log2FoldChange)
res_humoral[order_indices, ]
```
Visualisation for the first gene
```{r}
#plotCounts(dds, gene=which.max(res_humoral$log2FoldChange), intgroup="condition")
plotCounts(dds, gene=which.min(res_humoral$padj), intgroup="condition")
#plotCounts(dds, gene=rownames(res)[which.min(res$padj[which.max(res$log2FoldChange)])], intgroup="condition")
```
**Volcano plot**
```{r}
plt <- EnhancedVolcano(res_humoral,
                lab = rownames(res_humoral),
                x = "log2FoldChange",
                y = "padj",
                pCutoff = 0.05,
                FCcutoff = 1,
                labSize = 3.0,
                boxedLabels = FALSE,
                col = c('black', '#CBD5E8', '#B3E2CD', '#FDCDAC'),
                colAlpha = 1,
                title = NULL,        
                subtitle = NULL) 

plt
ggsave("./pictures_transpl/transpl_VolcanoPlot_humoral.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```

**Plot a heatmap of diff expressed genes**
```{r}
res_sign_humoral <- subset(res_humoral, padj < 0.05 & !is.na(padj) & abs(log2FoldChange) > 1.0)
res_sign_humoral <- res_sign_humoral[order(res_sign_humoral$log2FoldChange, decreasing = TRUE), ]

sig_genes <- rownames(res_sign_humoral)  # Получаем имена генов, которые прошли фильтрацию

de_mat <- assay(rlt)[sig_genes, ] 
datamatrix <- t(scale(t(de_mat)))

annotation_col <- data.frame(condition = coldata$condition)
rownames(annotation_col) <- colnames(datamatrix)

# coldata_humoral <- coldata[coldata$type == 150, ]
# rownames(coldata_150) <- coldata_150$sample
# coldata_150

annotation_colors <- list(
  condition = c("no_complications" = "#FFCC00", "humoral" = "#3399FF"))

pheatmap(datamatrix,
         cluster_rows = TRUE, 
         show_rownames = TRUE, 
         cluster_cols = TRUE, 
         annotation_col = annotation_col,
         annotation_colors = annotation_colors,
         display_numbers = TRUE,
         legend = FALSE,
         fontsize = 15)  

# ggsave("./pictures_transpl/transpl_Heatmap of diff expressed genes_humoral.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```


### Анализ обогащения гуморальный ответ ###
```{r}
up_humoral <- res_sign_humoral %>% 
  as.data.frame() %>% 
  filter(log2FoldChange > 0)
down_humoral <- res_sign_humoral %>% 
  as.data.frame() %>% 
  filter(log2FoldChange < 0)
rownames(up_humoral)
rownames(down_humoral)
```
перевела вручную: https://mirgenedb.org/browse/hsa

	•	miRBase: https://www.mirbase.org/
	•	MirGeneDB: https://mirgenedb.org/
```{r}
mirna_names_down <- c("hsa-miR-133a-3p", "hsa-miR-133a-3p", "hsa-miR-208a-3p", "hsa-miR-499a-5p")
```

**Конвертация в MIMATID**
```{r}
converted_mirna_up <- miRNAVersionConvert(mirna_names_up)
converted_mirna_down <- miRNAVersionConvert(mirna_names_down)
converted_mirna_up
converted_mirna_down
```

**Запрос таргетов из базы multiMiR** 
```{r}
targets150_down <- unique(get_multimir(org = "hsa", mirna = converted_mirna_up$Accession, table = "validated")@data$target_symbol)
targets150_up <- unique(get_multimir(org = "hsa", mirna = converted_mirna_down$Accession, table = "validated")@data$target_symbol)
#writeLines(targets_down, "targets_down150_list.txt")
#writeLines(targets_up, "targets_up150_list.txt")
```

**Анализ обогащения из базы биологических процессов**
```{r}
#msig_go_bp <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
# targets_down <- readLines("targets_down150_list.txt")
# targets_up <- readLines("targets_up150_list.txt")

GO_enrich_down150_bp <- enrichGO(
  gene          = targets150_down,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP", 
  pAdjustMethod = "BH",
  qvalueCutoff  = 0.05
)

GO_enrich_up150_bp <- enrichGO(
  gene          = targets150_up,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP", 
  pAdjustMethod = "BH",
  qvalueCutoff  = 0.05
)
```

**Визуализация**
```{r}

p1 <- dotplot(GO_enrich_down150_bp, showCategory = 20) + 
  ggtitle("GO Enrichment for DOWNregulated targets") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 11)
  )

p2 <- dotplot(GO_enrich_up150_bp, showCategory = 20) + 
  ggtitle("GO Enrichment for UPregulated targets") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 11)
  )

p1 + p2  

combined_plot <- p1 + p2

ggsave("./pictures_transpl/transpl_GO_enrichment_dotplot_BP_type150.tiff", plot = combined_plot, width = 16, height = 10, dpi = 300)
```
```{r}
GO_enrich_DOWN150_BP <- enrichplot::pairwise_termsim(GO_enrich_down150_bp, method = "JC")

plt <- emapplot(GO_enrich_DOWN150_BP, 
         repel = TRUE,
         showCategory = 20) +
  ggtitle("Biological processes for DOWNregulated targets for vesicles 150") +
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text = element_text(size = 3)
    )    

plt

ggsave("./pictures_transpl/transpl_GO_enrichment_emapplot_BPdown_type150.tiff", plot = plt, width = 16, height = 10, dpi = 300)
```
```{r}
GO_enrich_UP150_BP <- enrichplot::pairwise_termsim(GO_enrich_up150_bp, method = "JC")

plt <- emapplot(GO_enrich_UP150_BP, 
         repel = TRUE,
         showCategory = 20) +
  ggtitle("Biological processes for UPregulated targets for vesicles 150") +
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text = element_text(size = 3)
    )   

plt

ggsave("./pictures_transpl/transpl_GO_enrichment_emapplot_BPup_type150.tiff", plot = plt, width = 16, height = 10, dpi = 300)
```
**Анализ обогащения из базы IMMUNESIGDB**
IMMUNESIGDB: Наборы генов, связанные с иммунной системой, включая иммуно-онкологию и другие аспекты иммунитета.
```{r}
msig_go_bp <- msigdbr(species = "Homo sapiens", category = "C7", subcategory = "IMMUNESIGDB")

GO_enrich_up150_immum <- enricher(gene = targets150_up, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])
GO_enrich_down150_immun <- enricher(gene = targets150_down, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])
```

**Визуализация IMMUNESIGDB**
```{r}

p1 <- dotplot(GO_enrich_up150_immum, showCategory = 20) + 
  ggtitle("GO Enrichment IMMUNESIGD for UPregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p2 <- dotplot(GO_enrich_down150_immun, showCategory = 20) + 
  ggtitle("GO Enrichment IMMUNESIGD for DOWNregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p1 + p2

combined_plot <- p1 + p2

ggsave("./pictures_transpl/transpl_GO_enrichment_dotplot_IMMUNESIGD_type150.tiff", plot = combined_plot, width = 16, height = 10, dpi = 300)
```


**Анализ обогащения из базы KEGG**
```{r}
msig_go_bp <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG")

GO_enrich_up150_KEGG <- enricher(gene = targets150_up, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])
GO_enrich_down150_KEGG <- enricher(gene = targets150_down, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])

p1 <- dotplot(GO_enrich_up150_KEGG, showCategory = 20) + 
  ggtitle("GO Enrichment KEGG for UPregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p2 <- dotplot(GO_enrich_down150_KEGG, showCategory = 20) + 
  ggtitle("GO Enrichment KEGG for DOWNregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p1 + p2

combined_plot <- p1 + p2

ggsave("./pictures_transpl/transpl_GO_enrichment_dotplot_KEGG_type150.tiff", plot = combined_plot, width = 16, height = 10, dpi = 300)
```

**Анализ обогащения из базы CP:WIKIPATHWAYS**
```{r}
msig_go_bp <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:WIKIPATHWAYS")

GO_enrich_up150_WIKI <- enricher(gene = targets150_up, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])
GO_enrich_down150_WIKI <- enricher(gene = targets150_down, TERM2GENE = msig_go_bp[, c("gs_name", "gene_symbol")])

p1 <- dotplot(GO_enrich_up150_WIKI, showCategory = 20) + 
  ggtitle("GO Enrichment WIKIPATHWAYS for UPregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p2 <- dotplot(GO_enrich_down150_WIKI, showCategory = 20) + 
  ggtitle("GO Enrichment WIKIPATHWAYS for DOWNregulated targets for vesicles 150") + 
  theme(
    plot.title = element_text(size = 12, face = "bold"),
    axis.text.y = element_text(size = 8)
  )

p1 + p2
combined_plot <- p1 + p2

ggsave("./pictures_transpl/transpl_GO_enrichment_dotplot_WIKIPATHWAY_type150.tiff", plot = combined_plot, width = 16, height = 10, dpi = 300)
```
## Только везикулы типа 16
```{r}
coldata_16 <- coldata[coldata$type == 16, ]
rownames(coldata_16) <- coldata_16$sample
coldata_16
```
```{r}
common_samples_16 <- intersect(colnames(counts), coldata_16$samples)

counts_16 <- counts[, c(counts$miRNA, common_samples)]  
counts_16 <- counts_16[, rownames(coldata_16)] #ранжирую по колонки в counts так же как и названия строк в coldata_150
head(counts_16)
```
**Создаем DESeqDataSet из матрицы каунтов**
```{r}
dds_16 <- DESeqDataSetFromMatrix(countData = counts_16, 
                              colData = coldata_16, 
                              design = ~ 0 + patient + condition)
dds_16$condition <- relevel(dds_16$condition, ref = "before")
dds_16
```
**Фильтрация**
```{r}
dim(dds_16)
smallestGroupSize <- 3
keep <- rowSums(counts(dds_16) >= 10) >= smallestGroupSize
dds_16 <- dds_16[keep,]
dim(dds_16)
```

**Run Differential Expression Analysis for 150 type**
```{r}
dds_16 <- DESeq(dds_16, fitType = "parametric")
dds_16
```

```{r}
plotDispEsts(dds_16)
```
```{r}
res_16 <- results(dds_16, contrast=c("condition", "before", "after"))
res_16
```
**MA plot**

Фильтрация точек с низким средним экспрессированием (по baseMean).
	•	Обычно отсекаются baseMean < 1.
	2.	Определение значимых генов (синие точки):
	•	Используется критерий padj < 0.1 по умолчанию, а не < 0.05! 
```{r}
tiff("./pictures/PlotMA_standart_padj_0.05_type16.tiff", 
     width = 8, height = 6, units = "in", res = 300, bg = "white")
plotMA(res_16, alpha = 0.05, ylim = c(-8, 8)) 
dev.off()
plotMA(res_16, alpha = 0.05, ylim = c(-8, 8)) 
```
**Кастомный MA plot по p-value**
```{r}
res_df <- res_16 %>%
  as.data.frame %>%
  mutate(color = case_when( 
    pvalue < 0.05 & !is.na(pvalue) & abs(log2FoldChange) > 1  ~ "blue",  # Значимые по p-value и диф экспрессированные
    TRUE ~ "gray70"
  ))

plt <- ggplot(res_df, aes(x = baseMean, y = log2FoldChange, color = color)) +
  geom_point(alpha = 0.7, size = 1) +
  geom_hline(yintercept = 0, linetype = "solid", color = "gray40", size = 1.5) +  # Добавляем линию
  scale_color_manual(values = c("gray70" = "gray70", "blue" = "blue")) +
  scale_x_log10(labels = scales::scientific) + 
  theme_minimal() +
  labs(x = "mean of normalized counts", y = "log fold change") +
  theme(legend.position = "none")

plt
ggsave("./pictures_transpl/transpl_PlotMA_castom_pvalue0.05_type16.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
**Значимые результаты**
```{r}
summary(results(dds_16, contrast=c("condition", "before", "after"), alpha=0.05))
```
Let's arranged it by log2FoldChange:
```{r}
order_indices <- order(-res_16$log2FoldChange)
res_16[order_indices, ]
```
Visualisation for the first gene
```{r}
plotCounts(dds_16, gene=which.max(res_16$log2FoldChange), intgroup="condition")
plotCounts(dds_16, gene=which.min(res_16$pvalue), intgroup="condition")
#plotCounts(dds, gene=rownames(res)[which.min(res$padj[which.max(res$log2FoldChange)])], intgroup="condition")
```
**Volcano plot**
```{r}
plt <- EnhancedVolcano(res_16,
                lab = rownames(res_16),
                x = "log2FoldChange",
                y = "pvalue",
                pCutoff = 0.05,
                FCcutoff = 1,
                labSize = 3.0,
                boxedLabels = FALSE,
                col = c('black', '#CBD5E8', '#B3E2CD', '#FDCDAC'),
                colAlpha = 1,
                title = NULL,        
                subtitle = NULL) 

plt
ggsave("./pictures_transpl/transpl_Volcano plot_based_on_Pvalue_16type.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
**rlog трансформация**
	•	Черные точки – стандартное отклонение отдельных генов.
	•	Красная линия – сглаженный тренд зависимости SD от среднего значения экспрессии.
	•	Если красная линия наклонена вверх → стандартное отклонение растёт с увеличением среднего (плохая нормализация).
	•	Если красная линия примерно горизонтальна → нормализация сработала хорошо.
```{r}
rlt_16 <- rlog(dds_16) 
meanSdPlot(assay(rlt_16))  #показывает, как изменяется стандартное отклонение в зависимости от среднего значения экспрессии.
```
```{r}
pcaData <- plotPCA(rlt_16, intgroup=c("condition", "patient"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, shape = patient, color = condition)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "%")) +
  ylab(paste0("PC2: ", percentVar[2], "%")) + 
  coord_fixed() +
  theme_bw() +
  scale_color_brewer(palette = "Set2")
```
```{r}
assay(rlt_16) <- limma::removeBatchEffect(assay(rlt_16),
                                       batch = colData(dds_16)[,'patient'])

pcaData <- plotPCA(rlt_16, intgroup=c("condition", "patient"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

plt <- ggplot(pcaData, aes(PC1, PC2, shape = patient, color = condition)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "%")) +
  ylab(paste0("PC2: ", percentVar[2], "%")) + 
  coord_fixed() +
  theme_bw() +
  scale_color_brewer(palette = "Set2")

plt
ggsave("./pictures_transpl/transpl_PCA plot for type 16 after removing donor effect.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
**Plot a heatmap of the most expressed genes**
```{r}
res_sign_16 <- subset(res_16, pvalue < 0.05 & !is.na(pvalue) & abs(log2FoldChange) > 1.0)
res_sign_16 <- res_sign_16[order(res_sign_16$log2FoldChange, decreasing = TRUE), ]

sig_genes <- rownames(res_sign_16)  # Получаем имена генов, которые прошли фильтрацию

de_mat <- assay(rlt_16)[sig_genes, ] 
datamatrix <- t(scale(t(de_mat)))

annotation_col <- data.frame(condition = coldata_16$condition)
rownames(annotation_col) <- colnames(datamatrix)

annotation_colors <- list(
  condition = c("before" = "#FFCC00", "after" = "#3399FF")
)

plt <- pheatmap(datamatrix, 
         cluster_rows = TRUE, 
         show_rownames = TRUE, 
         cluster_cols = TRUE, 
         annotation_col = annotation_col,
         annotation_colors = annotation_colors,
         display_numbers = TRUE,
         legend = FALSE,
         fontsize = 15)  

plt
ggsave("./pictures_transpl/transpl_Heatmap of diff expressed genes_type16.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
**Plot of the distance between samples heatmap**
Расчет расстояний между образцами
	•	Обычно используется евклидово расстояние (по умолчанию в DESeq2).
	•	Оно вычисляется по нормализованным данным экспрессии (rlog() или vst()).
	•	Чем меньше расстояние — тем более похожи образцы.

```{r}
sampleDists_16 <- dist(t(assay(rlt_16)))
sampleDistMatrix_16 <- as.matrix(sampleDists_16)
rownames(sampleDistMatrix_16) <- paste(rlt_16$condition, rlt_16$patient, sep="_patient")
colnames(sampleDistMatrix_16) <- paste(rlt_16$condition, rlt_16$patient, sep="_patient")

plt <- pheatmap(sampleDistMatrix_16,
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         fontsize = 12,
         legend = FALSE,
         display_numbers = TRUE,
         color = colors)

plt
ggsave("./pictures_transpl/transpl_Plot of the distance between samples_type16.tiff", plot = plt, width = 8, height = 6, dpi = 300, bg = "white")
```
### Анализ обогащения для везикул типа 16 ###
```{r}
up_16 <- res_sign_16 %>% 
  as.data.frame() %>% 
  filter(log2FoldChange > 0)
down_16 <- res_sign_16 %>% 
  as.data.frame() %>% 
  filter(log2FoldChange < 0)
rownames(up_16)
rownames(down_16)
```
Переводим в miRBase
	•	miRBase: https://www.mirbase.org/
	•	MirGeneDB: https://mirgenedb.org/
```{r}
url <- "https://mirgenedb.org/browse/hsa"
page <- read_html(url)
```

Парсим таблицу
```{r}
mir_table <- page %>%
  html_element("table") %>%
  html_table(fill = TRUE) 

mir_table <- mir_table[-c(1:3), c(1,2) ] 
colnames(mir_table) <- c("MirGeneDB_ID", "MiRBase_ID")
mir_table$MirGeneDB_ID <- sub(" V", "", mir_table$MirGeneDB_ID)

head(mir_table)
```

```{r}
up_16_clean <- sub("_.*", "", row.names(up_16))
up_16_converted <- mir_table$MiRBase_ID[match(up_16_clean, mir_table$MirGeneDB_ID)]

#down_16_clean <- sub("_.*", "", row.names(down_16))
down_16_converted <- mir_table$MiRBase_ID[match(down_16_clean, mir_table$MirGeneDB_ID)]
up_16_converted
down_16_converted
```

**Конвертация в MIMATID** 
в итоге заменила NA вручную на самые близкие, но это такая себе практика

NA Hsa-Mir-10-P3a_5p  есть два соответствия:
Hsa-Mir-10-P1c =	hsa-mir-10a
Hsa-Mir-10-P3b = hsa-mir-125a

NA Hsa-Mir-126_3p* есть одно соответствие: 
Hsa-Mir-126-P2 = hsa-mir-126 

NA Hsa-Mir-30-P1c_3p* есть три соответствия:
Hsa-Mir-30-P1a = hsa-mir-30d
Hsa-Mir-30-P1b = hsa-mir-30a
Hsa-Mir-30-P1d = hsa-mir-30e

[1] "Hsa-Mir-185_5p"     "Hsa-Mir-197_3p"     "Hsa-Mir-10-P3a_5p"  "Hsa-Mir-148-P3_5p*" "Hsa-Mir-30-P1a_3p*" "Hsa-Mir-150_3p*"   
[7] "Hsa-Mir-361_3p*"    "Hsa-Mir-10-P2b_5p"  "Hsa-Mir-126_3p*"   
[1] "Hsa-Mir-340_5p"     "Hsa-Mir-181-P2c_5p" "Hsa-Mir-223_5p*"    "Hsa-Mir-30-P1c_3p*"

MI (MicroRNA Gene ID) — это идентификатор предшественника (precursor) miRNA
MIMAT (Mature miRNA ID) — это идентификатор зрелой (mature) miRNA, которая функционирует в клетке
```{r}
up_16_converted <-  c("hsa-mir-185-5p", "hsa-mir-197-3p", "hsa-mir-152-5p", "hsa-mir-30d-3p", "hsa-mir-150-3p", "hsa-mir-361-3p", "hsa-mir-99b-5p", "hsa-mir-126-3p")
down_16_converted <- c("hsa-mir-340-5p", "hsa-mir-181d-5p", "hsa-mir-223-5p")

# up_16_converted <-  c("hsa-mir-185", "hsa-mir-197", "hsa-mir-152", "hsa-mir-30d", "hsa-mir-150", "hsa-mir-361", "hsa-mir-99b", "hsa-mir-126")
# down_16_converted <- c("hsa-mir-340", "hsa-mir-181d", "hsa-mir-223") #но есть только MI, а не MIMAT

converted_mirna_up16 <- miRNAVersionConvert(up_16_converted)
converted_mirna_down16 <- miRNAVersionConvert(down_16_converted)
converted_mirna_up16
converted_mirna_down16
```
**Запрос таргетов из базы multiMiR**
```{r}
targets16_up <- unique(get_multimir(org = "hsa", mirna = converted_mirna_up16$Accession, table = "predicted")@data$target_symbol)
targets16_up
```
```{r}
targets16_down <- unique(get_multimir(org = "hsa", mirna = converted_mirna_down16$Accession, table = "predicted")@data$target_symbol)
targets16_down
```


